src/JustSaying/Messaging/Channels/IDownloadBuffer.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Amazon.SQS.Model;
+using JustSaying.AwsTools.MessageHandling;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal interface IDownloadBuffer
+    {
+        Task Start(CancellationToken stoppingToken);
+        ChannelReader<IQueueMessageContext> Reader { get; }
+    }
+
+    internal class DownloadBuffer : IDownloadBuffer
+    {
+        private readonly Channel<IQueueMessageContext> _channel;
+        private readonly int _bufferLength;
+        private readonly ISqsQueue _sqsQueue;
+        private readonly ILogger _logger;
+
+        // todo: add logic around populating this
+        private readonly List<string> _requestMessageAttributeNames = new List<string>();
+
+        public ChannelReader<IQueueMessageContext> Reader => _channel.Reader;
+
+        public DownloadBuffer(
+            int bufferLength,
+            ISqsQueue sqsQueue,
+            ILoggerFactory logger)
+        {
+            _bufferLength = bufferLength;
+            _sqsQueue = sqsQueue;
+            _logger = logger.CreateLogger<IDownloadBuffer>();
+            _channel = Channel.CreateBounded<IQueueMessageContext>(bufferLength);
+        }
+
+        public async Task Start(CancellationToken stoppingToken)
+        {
+            ChannelWriter<IQueueMessageContext> writer = _channel.Writer;
+            try
+            {
+                while (!stoppingToken.IsCancellationRequested)
```

Comment:
Something I haven't quite figured out yet - should use `.ThrowIfCancellationRequested()` rather than checking `. IsCancellationRequested` so that we always get an exception rather than it depending when it was cancelled?

Reply:
Fair point - this would probably make things a bit more consistent. In all my reading of the docs/internet I haven't really found a good explanation of when to use either, other than regular exception semantics - to short circuit continuations and ensure things are stopped _now_. Perhaps here we want to be a bit more graceful and notify consumers of completion instead of throwing up the stack?

Reply:
So generally, IMO, you should use ThrowIfCancellationRequested to signal you are terminating, and IsCancellationRequested if you want to detect the request for you to cancel, but you want to do some clean up before you ThrowIfCancellationRequested.

See item 1 here: https://devblogs.microsoft.com/premier-developer/recommended-patterns-for-cancellationtoken/

Reply:
Thanks, that's a great resource. Will change this.

Reply:
We've now reworked the cancellation model and tests to ensure we always throw on cancellation from each component.